src/test/node-modules-layout-maker_test.ts

```diff
@@ -0,0 +1,303 @@
+      a: {1: {c: '1'}},
+      c: {1: {b: '2'}},
+    };
+    // ROOT
+    // ├── A1
+    // ├── B1
+    // └── C1
+    //     └── B2
+    const expected: NodeModulesDirectory = {
+      a: {version: '1', nodeModules: {}},
+      b: {version: '1', nodeModules: {}},
+      c: {
+        version: '1',
+        nodeModules: {
+          b: {
+            version: '2',
+            nodeModules: {},
+          },
+        },
+      },
+    };
+    checkLayout(rootDeps, depGraph, expected);
+  });
+
+  test('version conflict between two branches', () => {
+    //     ROOT
+    //      /\
+    //     /  \
+    //    v    v
+    //   A1    B1
+    //   |      |
+    //   v      v
+    //  C1      C2
+    const rootDeps: PackageDependencies = {
+      a: '1',
+      b: '1',
+    };
+    const depGraph: DependencyGraph = {
+      a: {1: {c: '1'}},
+      b: {1: {c: '2'}},
+    };
+    // ROOT
+    // ├── A1
+    // ├── B1
+    // │   └── C2
+    // └── C1
+    const expected: NodeModulesDirectory = {
+      a: {version: '1', nodeModules: {}},
+      b: {
+        version: '1',
+        nodeModules: {
+          c: {version: '2', nodeModules: {}},
+        },
+      },
+      c: {version: '1', nodeModules: {}},
+    };
+    checkLayout(rootDeps, depGraph, expected);
+  });
+
+  test('version conflict requiring directory duplication', () => {
+    //      ROOT
+    //       /|\
+    //      / | \
+    //     /  |  \
+    //    v   v   v
+    //    A1  B1  C1
+    //         \   /
+    //          \ /
+    //           v
+    //           A2 --> D1
+    const rootDeps: PackageDependencies = {
+      a: '1',
+      b: '1',
+      c: '1',
+    };
+    const depGraph: DependencyGraph = {
+      a: {2: {d: '1'}},
+      b: {1: {a: '2'}},
+      c: {1: {a: '2'}},
+    };
+    // ROOT
+    // ├── A1
+    // ├── B1
+    // │   └── A2
+    // ├── C1
+    // │   └── A2
+    // └── D1
+    const expected: NodeModulesDirectory = {
+      a: {
+        version: '1',
+        nodeModules: {},
+      },
+      b: {
+        version: '1',
+        nodeModules: {
+          a: {version: '2', nodeModules: {}},
+        },
+      },
+      c: {
+        version: '1',
+        nodeModules: {
+          a: {version: '2', nodeModules: {}},
+        },
+      },
+      d: {
+        version: '1',
+        nodeModules: {},
+      },
+    };
+    checkLayout(rootDeps, depGraph, expected);
+  });
+
+  test('short loop', () => {
+    //   ROOT --> A1 --> B1
+    //             ^     |
+    //             |     |
+    //             +-----+
+    const rootDeps: PackageDependencies = {
+      a: '1',
+    };
+    const depGraph: DependencyGraph = {
+      a: {1: {b: '1'}},
+      b: {1: {a: '1'}},
+    };
+    // ROOT
+    // ├── A1
+    // └── B1
+    const expected: NodeModulesDirectory = {
+      a: {version: '1', nodeModules: {}},
+      b: {version: '1', nodeModules: {}},
+    };
+    checkLayout(rootDeps, depGraph, expected);
+  });
+
+  test('longer loop on a branch', () => {
```

Comment:
Yep, just did them manually.

For the file tree diagrams, I ran the `tree` command in some arbitrary directory on my computer so that I could easily copy-paste the ├ and └ characters.